cedar-policy-validator/src/custom_schema/to_json_schema.rs

```diff
@@ -0,0 +1,372 @@
+}
+
+pub trait TryInto<T>: Sized {
+    type Error;
+
+    fn try_into(self, context: &Context) -> Result<T, Self::Error>;
+}
+
+impl<T, U> TryInto<U> for T
+where
+    U: TryFrom<T>,
+{
+    type Error = U::Error;
+
+    fn try_into(self, context: &Context) -> Result<U, U::Error> {
+        U::try_from(self, context)
+    }
+}
+
+impl TryFrom<Type> for SchemaType {
+    type Error = ToValidatorSchemaError;
+    fn try_from(value: Type, context: &Context) -> Result<Self, Self::Error> {
+        Ok(match value {
+            Type::Ident(id) => match context.lookup(&id) {
+                Some(ty) => ty.into(),
+                None => Err(ToValidatorSchemaError::InvalidTypeName(id.to_smolstr()))?,
+            },
+            Type::Record(attrs) => Self::Type(TryInto::try_into(attrs, context)?),
+            Type::Set(b) => Self::Type(SchemaTypeVariant::Set {
+                element: Box::new(TryInto::try_into(b.node, context)?),
+            }),
+        })
+    }
+}
+
+impl TryFrom<EntityDecl> for EntityType {
+    type Error = ToValidatorSchemaError;
+    fn try_from(value: EntityDecl, context: &Context) -> Result<Self, Self::Error> {
+        Ok(Self {
+            member_of_types: value.member_of_types.map_or(vec![], |ns| {
+                ns.into_iter().map(|n| n.node.to_smolstr()).collect()
+            }),
+            shape: AttributesOrContext(SchemaType::Type(TryInto::try_into(value.attrs, context)?)),
+        })
+    }
+}
+
+impl TryFrom<Vec<Node<AttrDecl>>> for SchemaTypeVariant {
+    type Error = ToValidatorSchemaError;
+    fn try_from(value: Vec<Node<AttrDecl>>, context: &Context) -> Result<Self, Self::Error> {
+        let mut attrs: HashMap<Str, crate::TypeOfAttribute> = HashMap::new();
+        for n in value {
+            let attr_decl = n.node;
+            let name_str = name_to_str(attr_decl.name);
+            if let Some((ns, _)) = attrs.get_key_value(&name_str) {
+                return Err(ToValidatorSchemaError::DuplicateKeys(name_str, ns.clone()));
+            }
+            attrs.insert(
+                name_str,
+                TypeOfAttribute {
+                    ty: TryInto::try_into(attr_decl.ty.node, context)?,
+                    required: attr_decl.required.is_none(),
+                },
+            );
+        }
+        Ok(Self::Record {
+            attributes: attrs.into_iter().map(|(n, ty)| (n.node, ty)).collect(),
+            additional_attributes: false,
+        })
+    }
+}
+
+impl TryFrom<NonEmpty<Node<AppDecl>>> for ApplySpec {
+    type Error = ToValidatorSchemaError;
+    fn try_from(value: NonEmpty<Node<AppDecl>>, context: &Context) -> Result<Self, Self::Error> {
+        let mut resource_types = Vec::new();
+        let mut principal_types = Vec::new();
+        let mut attr_context: Option<AttributesOrContext> = None;
+        for decl in value {
+            match decl.node {
+                AppDecl::PR(PRAppDecl { ty, entity_tys }) => match ty.node {
```

Comment:
Is it ever possible that either arm of this match is hit more than once? 

If not, the way this is written in a loop is a bit confusing. Might be nicer if you could have a function `value.pr_app_decl` and `value.context`.